### DISCUSSION TITLE
How to guarantee cancel safety in custom method?

### DISCUSSION THREAD

**Discussion by lz1998** [tokio-rs/tokio#4416] [Q&A]
I want to define a method `read_packet()`, and use it in `tokio::select!`, but sometimes I get a negative length.

I have tried `FramedRead`, but it also doesn't work in `tokio::select!`.

```rust
    pub async fn read_packet(read_half: &mut OwnedReadHalf) -> Bytes {
        let len = read_half.read_i32().await.unwrap();
        if len - 4 < 0 {
            panic!("invalid packet length: {}", len);
        }
        let mut data = vec![0; len as usize - 4];
        read_half.read_exact(&mut data).await.unwrap();
        Bytes::from(data)
    }
```

**Comment by Darksonn** [comment]
Taken from the `tokio::select!` documentation:

> To determine whether your own methods are cancellation safe, look for the location of uses of .await. This is because when an asynchronous method is cancelled, that always happens at an .await. If your function behaves correctly even if it is restarted while waiting at an .await, then it is cancellation safe.

In your case, neither the `read_i32` or `read_exact` calls are cancellation safe, so anything built from them wont be either.

I would recommend using a codec from tokio-util. Unlike what you said in the question, the `FramedRead` type works great with a `tokio::select!` and its `next()` method is cancellation safe because it is a `Stream`, and `next()` is always cancellation safe on a `Stream`.
